[SYCL][Graph] Memory reuse for graph allocations in a single graph

Bensuo opened the pull request:
- Optimize memory use for allocations within a single graph by reusing memory where possible
- New handler impl member for node dependency access with the CGF
- New E2E tests for memory reuse
- Add missing CGType -> string conversion for graph printing alloc and free nodes

Bensuo commented on sycl/test-e2e/Graph/AsyncAlloc/Inputs/async_alloc_device_memory_reuse.cpp:
Hmm, it's a little tricky to test in a non-fragile way because of the alignment of the size to the device granularity that happens during the `async_mlloc` call.

In `async_alloc_queries.cpp` we can assume no memory reuse and are still only able to check like so:

```c++
  // Memory allocated might be adjusted for example based on device granularity,
  // so it may be more than expected but never less.
  assert(GraphExec.get_required_mem_size() >= ExpectedMinSize);
``` 

We could consider `ExpectedMinSize` for the reuse test to be the user passed size of one allocation, but given that most devices will have a larger granularity I don't think this would give us anything useful. The check would easily still pass if we broke memory reuse.

EwanC replied:
cool, let's leave it then.

EwanC commented on sycl/source/detail/async_alloc.cpp:
```diff
@@ -95,24 +110,38 @@ __SYCL_EXPORT void *async_malloc_from_pool(sycl::handler &h, size_t size,
+  // Get CG event dependencies for this allocation.
+  const auto &DepEvents = h.impl->CGData.MEvents;
+  auto UREvents = getUrEvents(DepEvents);
+  uint32_t numEvents = DepEvents.size();
```
Think we only need this on line 143, i'd just define this variable next to that usage, or inline it.

EwanC commented on sycl/source/queue.cpp:
```diff
@@ -440,9 +440,11 @@ sycl::detail::optional<event> queue::ext_oneapi_get_last_event_impl() const {
+  // the original event is used for tracking nodes in the graph.
   auto LastEventImpl = detail::getSyclObjImpl(*LastEvent);
-  if (LastEventImpl->isDiscarded() || LastEventImpl->isNOP())
+  if (!LastEventImpl->hasCommandGraph() &&
```
Is this `ext_oneapi_get_last_event` extension implementation something you found accidentally we should fix? wondering how it relates to async allocations.

Bensuo replied:
The async allocs that take a queue explicitly call `ext_oneapi_get_last_event()`:

```c++
__SYCL_EXPORT void *async_malloc(const sycl::queue &q, sycl::usm::alloc kind,
                                 size_t size,
                                 const sycl::detail::code_location &CodeLoc) {
  void *temp = nullptr;
  submit(
      q,
      [&](sycl::handler &h) {
        // In order queues must wait on the previous event before calling alloc.
        if (q.is_in_order() && q.ext_oneapi_get_last_event())
          h.depends_on(q.ext_oneapi_get_last_event().value());
        temp = async_malloc(h, kind, size);
      },
      CodeLoc);
  return temp;
}
```

Probably this would generally cause problems for a graph recording if used in user code also. I don't know that there's anything we need to fix necessarily, wrapping the event with another event causes the issue since we need the original event for association with a node. Perhaps it's worth looking at that as part of a greater examination of how we are using events throughout the graph, but I think that's out of scope for this work.

EwanC replied:
Thanks for the explanation, I'll create an internal ticket for us to specify the interaction of  extension `sycl_ext_oneapi_in_order_queue_events` with sycl-graph. There's also a `quue::ext_oneapi_set_external_event()` API as well as `queue::ext_oneapi_get_last_event()`, which I didn't know about until looking at it now.

fabiomestre commented on sycl/source/detail/graph_memory_pool.cpp:
```diff
@@ -0,0 +1,189 @@
+      continue;
+    }
+
+    // Check if the node is a free node, and if so is a free node for any of the
```
```suggestion
    // Check if the node is a free node and, if so, check if it is a free node for any of the
```

fabiomestre commented on sycl/source/detail/graph_memory_pool.cpp:
```diff
@@ -0,0 +1,189 @@
+
+    // Check if the node is a free node, and if so is a free node for any of the
+    // allocations which are free for reuse. We should not bother checking nodes
+    // that are free nodes, so we continue and check their predecessors.
```
```suggestion
    // that are not free nodes, so we continue and check their predecessors.
```

fabiomestre commented on sycl/source/detail/graph_memory_pool.cpp:
```diff
@@ -0,0 +1,189 @@
+    return false;
+  };
+
+  while (!NodesToCheck.empty()) {
```
Can we avoid the graph traversal when `CompatibleAllocs` is empty?

Bensuo replied:
I've added an early return for this case

fabiomestre commented on sycl/source/detail/async_alloc.cpp:
```diff
@@ -46,21 +46,36 @@ void *async_malloc(sycl::handler &h, sycl::usm::alloc kind, size_t size) {
   if (auto Graph = h.getCommandGraph(); Graph) {
-    alloc = Graph->getMemPool().malloc(size, kind);
+    // Get dependent graph nodes from any events
+    auto DepNodes = Graph->getNodesForEvents(DepEvents);
```
There seems to be some  code duplication. Would it make sense to share some of the code between async_malloc and async_malloc_from_pool()?

Bensuo replied:
Moved this to a separate helper function

Bensuo commented on sycl/source/detail/graph_memory_pool.cpp:
```diff
@@ -0,0 +1,189 @@
+  MGraph.resetNodeVisitedEdges();
+  // If we found an allocation, remove it from the free list.
+  if (AllocInfo) {
+    MFreeAllocations.erase(std::find(MFreeAllocations.begin(),
```
I'm a little torn on this. I have tested implementing a multimap instead and benchmarked it with 1000 allocations and 4 distinct key combinations. The benchmark is structured such that we're basically skipping the graph traversal and only testing looking up/collecting the free allocations. At best (on average) the multimap saves ~0.5us or around 10% of the overall time without graph traversal. At small numbers of nodes there is a lot of variance (even doing 10000 benchmark iterations) but it is often the same or slower with the multimap. 

I imagine that in a situation where graph traversal is necessary that will be where the most time is spent on a large graph.
IMO even at the large (possibly unrealistic for real usage) benchmark sizes I don't think this micro-optimization is worth the increased code complexity from working with a multimap.

fabiomestre replied:
I realize I said `multimap` but just to confirm whether you used `unordered_multimap` when benchmarking? I think that's probably the better option in this scenario.

I think that given the current focus on micro-optimizations in the sycl runtime, we should probably not be passing opportunities to optimize things even if the gains are small. I think it's important to have an implementation that scales well for large graph sizes as long as it doesn't cause issue for smaller graphs.

However,, it could be the case that the cost of creating a multimap with a complex key outweighs the benefits of using that container. If that's the case, I think we should just use an `std::list` and erase with an iterator. That still reduces the complexity of most operations without having to deal with multimaps.

Bensuo commented on sycl/source/detail/graph_memory_pool.cpp:
```diff
@@ -0,0 +1,189 @@
+
+  std::optional<alloc_info> AllocInfo = {};
+
+  // Called when traversing over nodes to check if the current node
```
Fixed thanks

Bensuo commented:
@intel/llvm-reviewers-runtime Could I get a review on this please? The Arc failures on CI look unrelated to this PR.

sarnex commented:
@Bensuo @EwanC 
This is causing a warning on GCC causing a build fail with -Werror (which we use)

```
FAILED: tools/sycl/source/CMakeFiles/sycl_object.dir/detail/graph_memory_pool.cpp.o 
ccache /usr/bin/g++ -DCL_TARGET_OPENCL_VERSION=300 -DSYCL2020_DISABLE_DEPRECATION_WARNINGS -DSYCL_EXT_JIT_ENABLE -DXPTI_ENABLE_INSTRUMENTATION -DXPTI_STATIC_LIBRARY -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__DPCPP_ENABLE_UNFINISHED_KHR_EXTENSIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__SYCL_INTERNAL_API -I/__w/llvm/llvm/build/tools/sycl/source -I/__w/llvm/llvm/src/sycl/source -I/__w/llvm/llvm/build/include -I/__w/llvm/llvm/src/xpti/include -I/__w/llvm/llvm/src/sycl/include -I/__w/llvm/llvm/build/_deps/boost_unordered-src/include -I/__w/llvm/llvm/build/_deps/boost_assert-src/include -I/__w/llvm/llvm/build/_deps/boost_config-src/include -I/__w/llvm/llvm/build/_deps/boost_container_hash-src/include -I/__w/llvm/llvm/build/_deps/boost_core-src/include -I/__w/llvm/llvm/build/_deps/boost_describe-src/include -I/__w/llvm/llvm/build/_deps/boost_mp11-src/include -I/__w/llvm/llvm/build/_deps/boost_predef-src/include -I/__w/llvm/llvm/build/_deps/boost_static_assert-src/include -I/__w/llvm/llvm/build/_deps/boost_throw_exception-src/include -I/__w/llvm/llvm/src/sycl-jit/common/include -I/__w/llvm/llvm/src/sycl-jit/jit-compiler/include -I/__w/llvm/llvm/src/unified-runtime/include -I/__w/llvm/llvm/build/_deps/opencl-headers-src -isystem /__w/llvm/llvm/src/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wextra -Werror -O3 -DNDEBUG -UNDEBUG -ffunction-sections -fdata-sections -fvisibility=hidden -fvisibility-inlines-hidden -Wno-psabi -std=c++17 -MD -MT tools/sycl/source/CMakeFiles/sycl_object.dir/detail/graph_memory_pool.cpp.o -MF tools/sycl/source/CMakeFiles/sycl_object.dir/detail/graph_memory_pool.cpp.o.d -o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/graph_memory_pool.cpp.o -c /__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp: In member function 'std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info> sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::tryReuseExistingAllocation(size_t, sycl::_V1::usm::alloc, bool, const std::vector<std::shared_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl> >&)':
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp:130:29: error: '((std::__weak_count<__gnu_cxx::_S_atomic>*)((char*)&AllocInfo + offsetof(std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>,std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::<unnamed>.std::_Optional_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false>::_M_payload.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false, false>::<unnamed>.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, true, false, false>::<unnamed>.std::_Optional_payload_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::_M_payload)))[7].std::__weak_count<__gnu_cxx::_S_atomic>::_M_pi' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  130 |   std::optional<alloc_info> AllocInfo = {};
      |                             ^~~~~~~~~
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp:130:29: error: '((std::__weak_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl, __gnu_cxx::_S_atomic>*)((char*)&AllocInfo + offsetof(std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>,std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::<unnamed>.std::_Optional_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false>::_M_payload.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false, false>::<unnamed>.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, true, false, false>::<unnamed>.std::_Optional_payload_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::_M_payload)))[3].std::__weak_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl, __gnu_cxx::_S_atomic>::_M_ptr' may be used uninitialized in this function [-Werror=maybe-uninitialized]
cc1plus: all warnings being treated as errors
[4160/5103] Building CXX object tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/scheduler.cpp.o
[4161/5103] Building CXX object tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o
[4162/5103] Building CXX object tools/sycl/source/CMakeFiles/sycl-preview_object.dir/detail/graph_memory_pool.cpp.o
FAILED: tools/sycl/source/CMakeFiles/sycl-preview_object.dir/detail/graph_memory_pool.cpp.o 
ccache /usr/bin/g++ -DCL_TARGET_OPENCL_VERSION=300 -DSYCL2020_DISABLE_DEPRECATION_WARNINGS -DSYCL_EXT_JIT_ENABLE -DXPTI_ENABLE_INSTRUMENTATION -DXPTI_STATIC_LIBRARY -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__DPCPP_ENABLE_UNFINISHED_KHR_EXTENSIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__SYCL_INTERNAL_API -I/__w/llvm/llvm/build/tools/sycl/source -I/__w/llvm/llvm/src/sycl/source -I/__w/llvm/llvm/build/include -I/__w/llvm/llvm/src/xpti/include -I/__w/llvm/llvm/src/sycl/include -I/__w/llvm/llvm/build/_deps/boost_unordered-src/include -I/__w/llvm/llvm/build/_deps/boost_assert-src/include -I/__w/llvm/llvm/build/_deps/boost_config-src/include -I/__w/llvm/llvm/build/_deps/boost_container_hash-src/include -I/__w/llvm/llvm/build/_deps/boost_core-src/include -I/__w/llvm/llvm/build/_deps/boost_describe-src/include -I/__w/llvm/llvm/build/_deps/boost_mp11-src/include -I/__w/llvm/llvm/build/_deps/boost_predef-src/include -I/__w/llvm/llvm/build/_deps/boost_static_assert-src/include -I/__w/llvm/llvm/build/_deps/boost_throw_exception-src/include -I/__w/llvm/llvm/src/sycl-jit/common/include -I/__w/llvm/llvm/src/sycl-jit/jit-compiler/include -I/__w/llvm/llvm/src/unified-runtime/include -I/__w/llvm/llvm/build/_deps/opencl-headers-src -isystem /__w/llvm/llvm/src/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wextra -Werror -O3 -DNDEBUG -UNDEBUG -D__INTEL_PREVIEW_BREAKING_CHANGES -ffunction-sections -fdata-sections -fvisibility=hidden -fvisibility-inlines-hidden -Wno-psabi -std=c++17 -MD -MT tools/sycl/source/CMakeFiles/sycl-preview_object.dir/detail/graph_memory_pool.cpp.o -MF tools/sycl/source/CMakeFiles/sycl-preview_object.dir/detail/graph_memory_pool.cpp.o.d -o tools/sycl/source/CMakeFiles/sycl-preview_object.dir/detail/graph_memory_pool.cpp.o -c /__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp: In member function 'std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info> sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::tryReuseExistingAllocation(size_t, sycl::_V1::usm::alloc, bool, const std::vector<std::shared_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl> >&)':
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp:130:29: error: '((std::__weak_count<__gnu_cxx::_S_atomic>*)((char*)&AllocInfo + offsetof(std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>,std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::<unnamed>.std::_Optional_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false>::_M_payload.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false, false>::<unnamed>.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, true, false, false>::<unnamed>.std::_Optional_payload_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::_M_payload)))[7].std::__weak_count<__gnu_cxx::_S_atomic>::_M_pi' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  130 |   std::optional<alloc_info> AllocInfo = {};
      |                             ^~~~~~~~~
/__w/llvm/llvm/src/sycl/source/detail/graph_memory_pool.cpp:130:29: error: '((std::__weak_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl, __gnu_cxx::_S_atomic>*)((char*)&AllocInfo + offsetof(std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>,std::optional<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::<unnamed>.std::_Optional_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false>::_M_payload.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, false, false, false>::<unnamed>.std::_Optional_payload<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info, true, false, false>::<unnamed>.std::_Optional_payload_base<sycl::_V1::ext::oneapi::experimental::detail::graph_mem_pool::alloc_info>::_M_payload)))[3].std::__weak_ptr<sycl::_V1::ext::oneapi::experimental::detail::node_impl, __gnu_cxx::_S_atomic>::_M_ptr' may be used uninitialized in this function [-Werror=maybe-uninitialized]
cc1plus: all warnings being treated as errors
```


To me it looks like a bogus warning but I'm not an expert in this code. Can someone take a look and work around the warning if it is bogus? Thanks

Bensuo commented:
> @Bensuo @EwanC This is causing a warning on GCC causing a build fail with -Werror (which we use)
> 
> https://github.com/intel/llvm/actions/runs/15128095339/job/42523770858
> 
> To me it looks like a bogus warning but I'm not an expert in this code. Can someone take a look and work around the warning if it is bogus? Thanks

I think this is bogus yeah, there is other similar initialization in the runtime that isn't getting flagged on GCC 11.4, and the error doesn't show up in 12.3. I've reworked the code so it's not initialized in the lambda (and fixed a few small issues I saw at the same time) and that seems fine now: https://github.com/intel/llvm/pull/18583

sarnex commented:
Thanks!
